### PR TITLE
part 1 

### PR DESCRIPTION
**Pull request by yotamgabi**

### CODE REVIEW

**Review by GuyMad232 — CHANGES_REQUESTED** [review]
layer_output = []
        for t in range(seq_len):
            prev_hidden_state = torch.stack(layer_states, dim=1)
            layer_states = []
            curr_x = input[:, t, :]

            for k, layer in enumerate(self.layer_params):
                h_k = prev_hidden_state[:, k, :]
                z = torch.sigmoid(layer['W_xz'](curr_x) + layer['W_hz'](h_k))
                r = torch.sigmoid(layer['W_xr'](curr_x) + layer['W_hr'](h_k))
                g = torch.tanh(layer['W_xn'](curr_x) + layer['W_hn'](r * h_k))
                h = z * h_k + (1 - z) * g
                layer_states.append(h.clone())
                curr_x = torch.nn.functional.dropout(h)

            layer_output.append(self.out_layer(curr_x))

        hidden_state = torch.stack(layer_states, dim=1)
        layer_output = torch.stack(layer_output, dim=1)
